src/Compilers/CSharp/Test/Semantic/Semantics/ReadOnlyStructsTests.cs

```diff
@@ -154,6 +154,400 @@ interface I1
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_PassFieldByRef()
+        {
+            var csharp = @"
+public struct S
+{
+    public static int f1;
+    public int f2;
+
+    public static void M1(ref int s) {}
+    public static void M2(in int s) {}
+
+    public void M3()
+    {
+        M1(ref f1); // ok
+        M1(ref f2); // ok
+        M2(in f1); // ok
+        M2(in f2); // ok
+    }
+
+    public readonly void M4()
+    {
+        M1(ref f1); // ok
+        M1(ref f2); // error
+        M2(in f1); // ok
+        M2(in f2); // ok
+    }
+}
+";
+            var comp = CreateCompilation(csharp);
+            // PROTOTYPE: should give ERR_RefReadonlyLocal
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_CallReadOnlyMethod()
+        {
+            var csharp = @"
+public struct S
+{
+    public int i;
+    public readonly int M1() => M2() + 1;
+    public readonly int M2() => i;
+}
+";
+            var comp = CompileAndVerify(csharp);
+            comp.VerifyIL("S.M1", @"
+{
+  // Code size        9 (0x9)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""int S.M2()""
+  IL_0006:  ldc.i4.1
+  IL_0007:  add
+  IL_0008:  ret
+}
+");
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void InParam_CallReadOnlyMethod()
+        {
+            var csharp = @"
+public struct S
+{
+    public int i;
+    public static int M1(in S s) => s.M2() + 1;
+    public readonly int M2() => i;
+}
+";
+            var comp = CompileAndVerify(csharp);
+            comp.VerifyIL("S.M1", @"
+{
+  // Code size        9 (0x9)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  call       ""int S.M2()""
+  IL_0006:  ldc.i4.1
+  IL_0007:  add
+  IL_0008:  ret
+}
+");
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_CallStaticMethod()
+        {
+            var csharp = @"
+public struct S
+{
+    public static int i;
+    public readonly int M1() => M2() + 1;
+    public static int M2() => i;
+}
+";
+            var comp = CompileAndVerify(csharp);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadOnlyMethod_CallNormalMethod()
+        {
+            var csharp = @"
+public struct S
+{
+    public int i;
+
+    public readonly void M1()
+    {
+        // should create local copy
+        M2();
+        System.Console.Write(i);
+
+        // explicit local copy, no warning
+        var copy = this;
+        copy.M2();
+        System.Console.Write(copy.i);
+    }
+
+    void M2()
+    {
+        i = 23;
+    }
+
+    static void Main()
+    {
+        var s = new S { i = 1 };
+        s.M1();
+    }
+}
+";
+
+            var verifier = CompileAndVerify(csharp, expectedOutput: "123");
+            // PROTOTYPE: should warn about copying 'this' when calling M2
+            verifier.VerifyDiagnostics();
+            verifier.VerifyIL("S.M1", @"
+{
```

Comment:
I think we should put these tests that are expected to produce IL in CodeGenReadOnlyStructTests

Reply:
Done